### PR TITLE
Add function gcd

### LINKED ISSUE #515: GCD (Greatest Common Divisior)
As suggested in #365, providing `gcd` function out of the box will be a good addition.

**Description**

Have a look at these comments to know more:
https://github.com/fortran-lang/stdlib/issues/365#issuecomment-907683782
https://github.com/j3-fortran/fortran_proposals/issues/221#issuecomment-907638974

**Prior Art**

Many languages provide `gcd` out of the box.

### PR DESCRIPTION
**Pull request by Sideboard**
Closes #515

1. Add elemental integer function gcd to calculate the greatest common divisor of two numbers
  a. Handle negative input as positive
2. Add tests (special, normal, swapped, negative, integer kinds)

### CODE REVIEW

**Comment by Sideboard** [comment]
I appreciate tips about tests. More, less, special cases …

**Comment by milancurcic** [comment]
Thanks @Sideboard, the implementation looks good to me. Tests seem fine as well. 

We should add the FORD docstrings to the function, and the spec as well. You can look at other functions for examples. Let us know if you need help with that.

General question for everybody, do we have a guide on when to put a function implementation in its own submodule and when to leave it in the module? I see that we have `clip` in the module but the others have their own submodules.

**Comment by Sideboard** [comment]
Ok, I can add them. I looked at clip for reference. Thought it was in the main file because of its size. The subfile question extends to the corresponding tests.

**Comment by Sideboard** [comment]
The spec only says "integer" and "another integer" for now.

File: doc/specs/stdlib_math.md
Comment (milancurcic) [inline doc/specs/stdlib_math.md:135]:
```suggestion
  a = 48
  b = 18
```

Kind suffix not necessary, compiler will cast it automatically.

Comment (Sideboard) [reply doc/specs/stdlib_math.md:135]: I took clip as a reference. Probably not the best choice.

Comment (Sideboard) [reply doc/specs/stdlib_math.md:135]: I would rather strip all the kinds for the example for simplicity. Or is it recommended to use explicit kinds in the examples? Still use a and b as input variables or simplify to use literals in the call?

Comment (milancurcic) [reply doc/specs/stdlib_math.md:135]: I suggest keeping `a` and `b` as declared variables. I would also just use default `integer` without `int32`. Makes the example simpler, and I can't think of a situation where it's problematic (i.e. default `integer` is `int32` on all platforms that I know of).

Comment (Sideboard) [reply doc/specs/stdlib_math.md:135]: Ok, I thought there was some underlying theme in the fact that clip uses int32 explicitly and didn't want to break them right next to it.

**Comment by aslozada** [comment]
> I appreciate tips about tests. More, less, special cases …

A test with nested functions, maybe.
`gcd(a,gcd(b,gcd(c,gcd(e,gcd(f,...gcd(y,x))))))`

**Comment by Sideboard** [comment]
> A test with nested functions, maybe.
> `gcd(a,gcd(b,gcd(c,gcd(e,gcd(f,...gcd(y,z))))))`

Does this really test anything new? If `gcd(y, z)` correctly returns `yz` what does `gcd(x, yz)` tell us?

Is there interest in a version of gcd that works with rank-1 arrays for more than two arguments?

**Comment by Carltoffel** [comment]
> Is there interest in a version of gcd that works with a rank-1 array for more than two arguments?

What do you mean? Could you give an example, please?

**Comment by Sideboard** [comment]
> Edit: Something like gcd([20, 30, 40]) -> 10?

Yes, that's what I mean.

**Comment by ivan-pi** [comment]
> General question for everybody, do we have a guide on when to put a function implementation in its own submodule and when to leave it in the module? I see that we have `clip` in the module but the others have their own submodules.

On the long term I think the public modules should contain only the interfaces, and all of the implementation should be placed in sub-modules. This would facilitate compiler vendors providing their own optimized implementations in the (distant?) future. It would also make it easier to swap between debug and release binaries by re-linking. I think the current reasoning was simply when launching a new module with a small function, it was just easier to prepare a single module.

A second reason in favor of separating interface and implementation *might* be to use a stdlib library compiled by another Fortran compiler. In this case you would only need to compile the interface (see https://github.com/fortran-lang/stdlib/pull/530#discussion_r711945701 and the discussion that follows). As an example, Intel MKL ships libraries that can be used with both Intel and gfortran compilers.

@milancurcic, should I make a new issue to discuss this? We can then edit the developer docs in a new PR.

**Comment by milancurcic** [comment]
Yes @ivan-pi, thanks, that would be good.

**Comment by Sideboard** [comment]
Should anything be changed for this PR?

File: src/stdlib_math.fypp
Comment (Aman-Godara) [inline src/stdlib_math.fypp:306]: This part needs to be moved up at `gcd` interface. You can write any comments here specific to this particular procedure like:
```suggestion
    !> Returns greatest common divisor of two integers of kind ${k1}$
```

File: doc/specs/stdlib_math.md
Comment (Aman-Godara) [inline doc/specs/stdlib_math.md:115]: This part needs to be improved a bit.
You can mention `intent` here. If possible try to make it similar to other procedures mentioned in this doc to be consistent.

File: doc/specs/stdlib_math.md
Comment (Aman-Godara) [inline doc/specs/stdlib_math.md:131]: Should `kind` be mentioned explicitly here to emphasize the point made in line 117?

File: doc/specs/stdlib_math.md
Comment (Aman-Godara) [inline doc/specs/stdlib_math.md:136]: I am trying to think like a beginner for whom it is the first programming language.

if `==` is accepted as a .true. statement then it is fine, I personally would have preferred `<--` (or any other random text) to indicate that the value of c is 6 because `==` operator, I think, is little ambiguous as it can return `.false.` also. Your call, shouldn't be a problem either way.

**Review by Aman-Godara — APPROVED** [review]
Looks great to me! I have left few comments, some of them are just open ended question for everyone. Thank you @Sideboard.

File: doc/specs/stdlib_math.md
Comment (milancurcic) [reply doc/specs/stdlib_math.md:136]: We could also write it something like this:

```
c = gcd(a, b) ! returns 6
```

File: src/stdlib_math.fypp
Comment (Sideboard) [reply src/stdlib_math.fypp:306]: I switched back and forth between "returns" and "determines", it's not uniform among the existing procedures as far as I remember.

Is there some guide about what goes where? I did not know if the doc scripts would recognize annotations to the interface declarations.

File: doc/specs/stdlib_math.md
Comment (Sideboard) [reply doc/specs/stdlib_math.md:115]: I am not sure what to write for the a and b. Don't want to repeat the description of the function.

I thought intent would be read from the code itself.

File: doc/specs/stdlib_math.md
Comment (Sideboard) [reply doc/specs/stdlib_math.md:131]: There was explicit kinds here initially but I was asked to remove them since int32 is the default kind.

File: doc/specs/stdlib_math.md
Comment (Sideboard) [reply doc/specs/stdlib_math.md:136]: I find arrows confusing. I like the return version.

File: src/stdlib_math.fypp
Comment (14NGiestas) [reply src/stdlib_math.fypp:306]: With "returns" it get's clear that this procedure is expected to return a value.

File: src/stdlib_math.fypp
Comment (14NGiestas) [inline src/stdlib_math.fypp:321]: If anyone ever wonder about performance in GCD there is a interesting benchmark
[here (in C)](https://hbfs.wordpress.com/2013/12/10/the-speed-of-gcd/). In conclusion it seems that this implementation is clear and fast enough.
PS: There is also the [gmplib's documentation](https://gmplib.org/manual/Greatest-Common-Divisor-Algorithms) that explores other GCD algorithms.

I believe this patch is good to go

**Review by 14NGiestas — APPROVED** [review]
There is only a minor (non-)issue about the wording in the docs but I believe this is good to go.

**Review by ivan-pi — APPROVED** [review]
Thanks @14NGiestas for the interesting link. There was a [nice response](https://lemire.me/blog/2013/12/26/fastest-way-to-compute-the-greatest-common-divisor/) to that blog post from Prof Daniel Lemire (@lemire). 

Pending the small issues in the documentation, I approve merging this.

**Comment by Sideboard** [comment]
Is there a consensus on where the annotation should be, the interface or the implementation?

**Comment by Sideboard** [comment]
Both descriptive "Converts …" and imperative "Return …" can be found among the function annotations.

**Review by 14NGiestas — APPROVED** [review]

**Review by Aman-Godara — APPROVED** [review]

**Comment by Aman-Godara** [comment]
@milancurcic this still needs one more approval to be merged.

File: src/tests/math/test_stdlib_math.f90
Comment (Sideboard) [reply src/tests/math/test_stdlib_math.f90]: I added (0, -2). Now I see that it's not exactly what was asked for but probably close enough.

**Comment by Sideboard** [comment]
I thought the approvals were invalidated when I pushed to the branch but a grey check mark means that the reviewer does not have write-access necessary for the merge. Should I request a review from someone with those rights? Who?

**Comment by milancurcic** [comment]
Thanks all, let's merge it.

**Pull request merged**
